src/coreclr/jit/morph.cpp

```diff
@@ -293,6 +293,39 @@ GenTree* Compiler::fgMorphExpandCast(GenTreeCast* tree)
     var_types dstType = tree->CastToType();
     unsigned  dstSize = genTypeSize(dstType);
 
+#if defined(TARGET_AMD64)
+    // If AVX512 is present, we have intrinsic available to convert
+    // ulong directly to float. Hence, we need to combine the 2 nodes
+    // GT_CAST(GT_CAST(TYP_ULONG, TYP_DOUBLE), TYP_FLOAT) into a single
+    // node i.e. GT_CAST(TYP_ULONG, TYP_FLOAT). At this point, we already
+    // have the 2 GT_CAST nodes in the tree and we are combining them below.
+    if (oper->OperIs(GT_CAST))
+    {
+        GenTreeCast* innerCast = oper->AsCast();
+
+        if (innerCast->IsUnsigned())
+        {
+            GenTree*  innerOper    = innerCast->CastOp();
+            var_types innerSrcType = genActualType(innerOper);
+            var_types innerDstType = innerCast->CastToType();
+            unsigned  innerDstSize = genTypeSize(innerDstType);
+            innerSrcType           = varTypeToUnsigned(innerSrcType);
+
+            // Check if we are going from ulong->double->float
+            if (innerSrcType == TYP_ULONG && innerDstType == TYP_DOUBLE && dstType == TYP_FLOAT)
+            {
+                if (compOpportunisticallyDependsOn(InstructionSet_AVX512F))
+                {
+                    // One optimized (combined) cast here
+                    tree         = gtNewCastNode(TYP_ULONG, innerOper, true, TYP_FLOAT);
```

Comment:
I think this should be `TYP_LONG` with `unsigned:true`. This works but it's sort of invalid IR, right @SingleAccretion ?

Reply:
This code does not make sense to me. Why is it creating a `TYP_ULONG` (invalid) node only to immediately reset its type to `TYP_FLOAT` (as was, I presume, intended)?

Reply:
ah, didn't notice that `tree->gtType = TYP_FLOAT;` - I agree, looks a bit weird

Reply:
@khushal1996, this should be a simple fix. It can likely just be `gtNewCast(TYP_FLOAT, innerOper, true, TYP_FLOAT)` directly.

Reply:
@tannergooding  Thanks for the suggestion. 

@EgorBo @SingleAccretion Thanks for the review. I was under the impression that the first argument was the input type. But then during debugging, I set it back to float for correctness. Thanks for the reminder. I have changed the code according to the suggestions.